src/rules/noAsyncWithoutAwaitRule.ts

```diff
@@ -0,0 +1,139 @@
+
+type FunctionNodeType = ts.ArrowFunction | ts.FunctionDeclaration | ts.MethodDeclaration | ts.FunctionExpression;
+
+export class Rule extends Lint.Rules.AbstractRule {
+    public static FAILURE_STRING = "Functions marked async must contain an await or return statement.";
+
+    public static metadata: Lint.IRuleMetadata = {
+        codeExamples,
+        description: Rule.FAILURE_STRING,
+        hasFix: false,
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        /* tslint:disable:max-line-length */
+        rationale: Lint.Utils.dedent`
+        Marking a function as \`async\` without using \`await\` or returning a value inside it can lead to an unintended promise return and a larger transpiled output.
+        Often the function can be synchronous and the \`async\` keyword is there by mistake.
+        Return statements are allowed as sometimes it is desirable to wrap the returned value in a Promise`,
+        /* tslint:enable:max-line-length */
+        ruleName: "no-async-without-await",
+        type: "functionality",
+        typescriptOnly: false,
+    };
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new Walk(sourceFile, this.getOptions()));
+    }
+}
+
+class Walk extends Lint.RuleWalker {
+    protected visitFunctionDeclaration(node: ts.FunctionDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionDeclaration(node);
+    }
+
+    protected visitArrowFunction(node: ts.ArrowFunction) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitArrowFunction(node);
+    }
+
+    protected visitMethodDeclaration(node: ts.MethodDeclaration) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitMethodDeclaration(node);
+    }
+
+    protected visitFunctionExpression(node: ts.FunctionExpression) {
+        this.addFailureIfAsyncFunctionHasNoAwait(node);
+        super.visitFunctionExpression(node);
+    }
+
+    private getAsyncModifier(node: ts.Node) {
+        if (node.modifiers !== undefined) {
+            return node.modifiers.find((modifier) => modifier.kind === ts.SyntaxKind.AsyncKeyword);
+        }
+
+        return undefined;
+    }
+
+    private isAwait(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.AwaitKeyword;
+    }
+
+    private isReturn(node: ts.Node): boolean {
+        return node.kind === ts.SyntaxKind.ReturnKeyword;
+    }
+
+    private functionBlockHasAwait(node: ts.Node): boolean {
+        if (this.isAwait(node)) {
+            return true;
+        }
+
+        if (node.kind === ts.SyntaxKind.ArrowFunction
+            || node.kind === ts.SyntaxKind.FunctionDeclaration) {
+            return false;
+        }
+
+        // tslint:disable-next-line:no-unsafe-any
+        return node.getChildren().some(this.functionBlockHasAwait.bind(this));
```

Comment:
> this.functionBlockHasAwait.bind(this)

`.bind` is still somewhat untyped. Instead of having to violate `no-unsafe-any`, separate `functionBlockHasAwait` into a standalone `function functionBlockHasAwait`. You'll have to move any other member functions that it calls to separate functions too.

Reply:
Ping @eranshabi - PR is pending this comment & some others.